python/utils/OpaqueArguments.h

```diff
@@ -156,115 +186,63 @@ packArgs(OpaqueArguments &argData, py::args args,
     auto kernelArgTy = kernelFuncOp.getArgument(i).getType();
     llvm::TypeSwitch<mlir::Type, void>(kernelArgTy)
         .Case([&](mlir::ComplexType ty) {
-          if (!py::hasattr(arg, "real"))
-            throw std::runtime_error("invalid complex element type");
-          if (!py::hasattr(arg, "imag"))
-            throw std::runtime_error("invalid complex element type");
-
-          std::complex<double> *ourAllocatedArg = new std::complex<double>();
+          checkArgumentType<py_ext::Complex>(arg, i);
           if (isa<Float64Type>(ty.getElementType())) {
-            *ourAllocatedArg =
-                std::complex<double>(PyFloat_AsDouble(arg.attr("real").ptr()),
-                                     PyFloat_AsDouble(arg.attr("imag").ptr()));
-
-            argData.emplace_back(ourAllocatedArg, [](void *ptr) {
-              delete static_cast<std::complex<double> *>(ptr);
-            });
+            auto *ourAllocatedArg =
```

Comment:
since you've created `addArgument<T>` can't you merge the `ourAllocatedArg` allocation into that new function as well and just have a one-liner (by also taking the `arg` as input)?

Reply:
There are differences in how we initialize the argument - sometimes we need to cast it but sometimes not... Let me see if I can add two functions here

Reply:
Done!